Build a regression model.

In [5]:
import requests
import json
import pandas as pd
import csv
import seaborn as sns
import matplotlib.pyplot as plt
import ast

Provide model output and an interpretation of the results. 

In [6]:
merged_df = pd.read_csv("../csv_folder/bikes_places.csv")


H0: All coefficients = 0 (Being near POIs has no effect on free_bikes).

H1: At least one coefficient != 0 (At least one of the POI categories has an impact on free_bikes).

In [ ]:
import statsmodels.api as sm

X = merged_df[['Bar/Restaurant','Bar', 'Restaurant','Park','Other']]  # INDEPENDENT VARIABLES
y = merged_df['free_bikes']  # DEPENDENT VARIABLES

X = sm.add_constant(X)  # INTERCEPT
model = sm.OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.068
Model:                            OLS   Adj. R-squared:                  0.063
Method:                 Least Squares   F-statistic:                     13.63
Date:                Tue, 29 Jul 2025   Prob (F-statistic):           7.45e-13
Time:                        10:43:41   Log-Likelihood:                -3115.4
No. Observations:                 937   AIC:                             6243.
Df Residuals:                     931   BIC:                             6272.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const              5.1707      6.058      0.

### BAR/RESUAURANT:
p = 0.69 -> Weak evidence but borderline at α = 0.05 

### PARK
p = 0.130 -> not significant

### BAR
p = 0.818 -> not significant

### RESTAURANT
p = 0.940 -> not significant

### OTHER
p = 0.798 -> not significant


### Based on the numbers above, NULL hypothesis is true 

## Since free_bikes is a discrete, not continious Poisson regression is the right model

In [ ]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

# DEFINE THE FORMILA WITH PREDICTORS
formula = 'free_bikes ~ Q("Bar/Restaurant") + Bar + Restaurant + Park + Other'

# FIT POISSON MODEL
poisson_model = smf.glm(formula=formula, data=merged_df, family=sm.families.Poisson()).fit()

print(poisson_model.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:             free_bikes   No. Observations:                  937
Model:                            GLM   Df Residuals:                      931
Model Family:                 Poisson   Df Model:                            5
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -4466.3
Date:                Tue, 29 Jul 2025   Deviance:                       6174.4
Time:                        10:43:44   Pearson chi2:                 6.41e+03
No. Iterations:                     5   Pseudo R-squ. (CS):             0.3663
Covariance Type:            nonrobust                                         
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Intercept               1.7461    

### Intercept = 1.7461 (p < 0.001)
When all category indicators are 0 (baseline group), the expected number of free bikes is ~5.7

### Bar/Restaurant 
Coefficient: 0.1688
p < 0.001 -> Strong evidence

## Bar
Coefficient: -0.0273
p = 0.447 -> Weak evidence

## Restaurant
Coefficient: -0.0134
p = 0.697 -> Wak evidence

## Park
Coefficient: 0.1153
P = 0.001 -> Strong evidence

## Other
Coefficient: -0.0367
P = 0.334 -> Weak evidence



### Based on the numbers above, we can reject the NULL hypothesis and we have Bar/Restaurant and Park categories  AS POIs that have an impact on free bikes





# Stretch

How can you turn the regression model into a classification model?